File: commons/hexagon/templates/kotlin/src/main/kotlin/com/phenotype/domain/ports/outbound/Repository.kt
package com.phenotype.domain.ports.outbound

import com.phenotype.domain.model.ExampleEntity

/**
 * Outbound port for repository
 */
interface Repository {
    /**
     * Save an entity
     */
    suspend fun save(entity: ExampleEntity)

    /**
     * Find entity by ID
     */
    suspend fun findById(id: String): ExampleEntity?

    /**
     * Find all entities
     */
    suspend fun findAll(): List<ExampleEntity>

    /**
     * Delete entity by ID
     */
    suspend fun delete(id: String)

    /**
     * Check if entity exists
     */
    suspend fun exists(id: String): Boolean
}

/**
 * Outbound port for unit of work
 */
interface UnitOfWork {
    val repository: Repository

    suspend fun commit()
    suspend fun rollback()
}

File: commons/templates/hexagonal/hexagonal-go/domain/ports/outbound/repository.go
package outbound

import (
	"context"

	"hexagonal-go/domain/entities"
	"hexagonal-go/domain/valueobjects"
)

// Repository defines the outbound port for entity persistence
type Repository interface {
	// Save persists an entity
	Save(ctx context.Context, entity *entities.Entity) error

	// FindByID retrieves an entity by ID
	FindByID(ctx context.Context, id string) (*entities.Entity, error)

	// Delete removes an entity
	Delete(ctx context.Context, id string) error

	// List returns all entities with pagination
	List(ctx context.Context, pagination valueobjects.Pagination) ([]*entities.Entity, error)
}
